### PR TITLE
Make services a required field in sbr spec

### PR DESCRIPTION
**Pull request by pratikjagrut**
### Motivation

https://issues.redhat.com/browse/APPSVC-696

### Testing

`make test-acceptance VERBOSE=3`

### CODE REVIEW

File: test/acceptance/features/bindAppToService.feature [hunk 1]
```diff
@@ -585,3 +585,33 @@ Feature: Bind an application to a service
         And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding "binding-request-backend-a" should be changed to "False"
         And Secret "binding-request-backend-a" contains "BACKEND_READY" key with value "true"
         And Secret "binding-request-backend-a" contains "BACKEND_HOST" key with value "example.common"
+
+        @negative
+        Scenario: Create Service Binding with empty services
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:590]: you cannot create it, as you show later on. Hence, the scenario should reflect that, something like:

"Service Binding with empty services is not allowed in the cluster"

File: test/acceptance/features/bindAppToService.feature [hunk 2]
```diff
@@ -585,3 +585,33 @@ Feature: Bind an application to a service
...
+        When Invalid Service Binding is applied
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:591]: We can reuse already existing step `Service Binding is applied`. What is the benefit of the new step?

File: test/acceptance/features/bindAppToService.feature [hunk 3]
```diff
@@ -585,3 +585,33 @@ Feature: Bind an application to a service
...
+            """
+            apiVersion: operators.coreos.com/v1alpha1
+            kind: ServiceBinding
+            metadata:
+                name: binding-request-empty-services
+            spec:
+                services:
+            """
+        Then Openshift throws an error "invalid: spec.services: Invalid value: \"null\""
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:600]: Let be more generic here, for future plain k8s support: how about "Cluster reports the error "invalid: spec.services: Invalid value: \"null\""

Also, we really need to check that the submitted SB instance is not persisted in the cluster

File: test/acceptance/features/bindAppToService.feature [hunk 4]
```diff
@@ -585,3 +585,33 @@ Feature: Bind an application to a service
...
+
+        @negative
+        Scenario: Create Service Binding without GVK of service
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:603]: see above.

**Review by pedjak — CHANGES_REQUESTED** [review]
in addition to other comments:

* please squash commits
* we need additional test case where `services` fields is not present in SB at all
* we need scenario where existing valid SB gets updated so that `services` fields are removed.

File: test/acceptance/features/steps/steps.py [hunk 5]
```diff
@@ -425,3 +426,17 @@ def etc_cluster_is_running(context, etcd_name):
...
         assert etcd_cluster.create() is True, "etcd cluster is not created"
         assert etcd_cluster.is_present() is True, "etcd cluster is not present"
+
+
+@when(u'Invalid Service Binding is applied')
+def invalid_sbr_is_applied(context):
+    yaml = context.text
+    sbr = ServiceBinding()
+    output = sbr.attempt_to_create(yaml)
+    context.error = output
```
Comment (pmacik) [inline test/acceptance/features/steps/steps.py:436]: Please rename the context variable to `context.expected_error` so that it is clear that it is expected instead of being confused with an ordinary error.

File: test/acceptance/features/steps/service_binding.py [hunk 6]
```diff
@@ -10,3 +10,6 @@ class ServiceBinding(object):
     def create(self, yaml):
         sbr_create_output = self.openshift.oc_apply(yaml)
```
Comment (pmacik) [inline test/acceptance/features/steps/service_binding.py:11]: Please, replace this line with
```go
sbr_create_output = self.attempt_to_create(yaml)
```
to avoid duplicating pieces of code.

File: test/acceptance/features/steps/steps.py [hunk 7]
```diff
@@ -425,3 +426,17 @@ def etc_cluster_is_running(context, etcd_name):
...
+
+
+@then(u'Openshift throws an error "{err_msg}"')
```
Comment (pmacik) [inline test/acceptance/features/steps/steps.py:439]: Let's be more generic here in the step definition... since we are actually asserting the content of `context.expected_error` here, the step should rather be something like `Error message "{err_msg}" is thrown` instead of `Openshift throws error...`. 

So please rename the stes to the `Error message "{err_msg}" is thrown` so that it can be re-used by other scenarios. Where the previous steps would set `context.expected_error` to be checked later with this step.

**Review by pmacik — CHANGES_REQUESTED** [review]

File: test/acceptance/features/bindAppToService.feature (hunk 3 above)
Comment (pmacik) [reply test/acceptance/features/bindAppToService.feature:600]: @pedjak We can be even more generic here, see https://github.com/redhat-developer/service-binding-operator/pull/698#discussion_r498217503

File: test/acceptance/features/bindAppToService.feature (hunk 2 above)
Comment (pmacik) [reply test/acceptance/features/bindAppToService.feature:591]: The new step has a different assertion expectations. The original step expect no error to be thrown, while the new step expect error to be thrown... so the behavior of each variant is different, therefore it should be split.

File: test/acceptance/features/bindAppToService.feature (hunk 3 above)
Comment (pmacik) [reply test/acceptance/features/bindAppToService.feature:600]: +1 on checking SB is not persisted.

File: test/acceptance/features/bindAppToService.feature (hunk 1 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:590]: Ahh, you got a point here, I'll update it.

Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature]: Done.

File: test/acceptance/features/bindAppToService.feature (hunk 4 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature]: Done.

File: test/acceptance/features/steps/steps.py (hunk 5 above)
Comment (pratikjagrut) [reply test/acceptance/features/steps/steps.py]: Done.

File: test/acceptance/features/steps/service_binding.py (hunk 6 above)
Comment (pratikjagrut) [reply test/acceptance/features/steps/service_binding.py]: Done.

File: test/acceptance/features/steps/steps.py (hunk 7 above)
Comment (pratikjagrut) [reply test/acceptance/features/steps/steps.py]: Done.

Comment (pratikjagrut) [reply test/acceptance/features/steps/steps.py]: PTAL

File: test/acceptance/features/bindAppToService.feature (hunk 3 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature]: I made some changes PTAL.

File: pkg/apis/operators/v1alpha1/servicebinding_types.go [hunk 8]
```diff
@@ -112,7 +113,8 @@ type ServiceBinding struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec   ServiceBindingSpec   `json:"spec,omitempty"`
+	// +required
+	Spec   ServiceBindingSpec   `json:"spec"`
```
Comment (pedjak) [inline pkg/apis/operators/v1alpha1/servicebinding_types.go:117]: is this change really necessary in this PR?

File: test/acceptance/features/bindAppToService.feature [hunk 9]
```diff
@@ -6,9 +6,8 @@ Feature: Bind an application to a service
     Background:
         Given Namespace [TEST_NAMESPACE] is used
         * Service Binding Operator is running
-        * PostgreSQL DB operator is installed
+        # * PostgreSQL DB operator is installed
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:9]: please remove if not needed.

File: test/acceptance/features/bindAppToService.feature [hunk 10]
```diff
@@ -586,3 +585,86 @@ Feature: Bind an application to a service
         And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding "binding-request-backend-a" should be changed to "False"
         And Secret "binding-request-backend-a" contains "BACKEND_READY" key with value "true"
         And Secret "binding-request-backend-a" contains "BACKEND_HOST" key with value "example.common"
+
+        @negative1
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:589]: `@negative1` is the new annotation?

File: test/acceptance/features/bindAppToService.feature [hunk 11]
```diff
@@ -586,3 +585,86 @@ Feature: Bind an application to a service
...
+            """
+            apiVersion: operators.coreos.com/v1alpha1
+            kind: ServiceBinding
+            metadata:
+                name: binding-request-without-spec
+            """
+        Then Error message "spec: Required value" is thrown
+        And Service Binding "binding-request-without-spec" is not persistent in the cluster
+
+        @negative1
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:601]: same as above

File: test/acceptance/features/bindAppToService.feature [hunk 12]
```diff
@@ -586,3 +585,86 @@ Feature: Bind an application to a service
...
+            kind: ServiceBinding
+            metadata:
+                name: binding-request-empty-services
+            spec:
+                services:
+            """
+        Then Error message "invalid: spec.services: Invalid value: \"null\"" is thrown
+        And Service Binding "binding-request-empty-services" is not persistent in the cluster
+
+        @negative1
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:628]: same as above

File: test/acceptance/features/bindAppToService.feature [hunk 13]
```diff
@@ -586,3 +585,86 @@ Feature: Bind an application to a service
...
+            spec:
+                services:
+                -   name: backend-demo
+            """
+        Then Error message "spec.services.group: Required value" is thrown
+        And Error message "spec.services.kind: Required value" is thrown
+        And Error message "spec.services.version: Required value" is thrown
+        And Service Binding "binding-request-without-gvk" is not persistent in the cluster
+
+        @negative1
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:645]: same as above

File: test/acceptance/features/bindAppToService.feature [hunk 14]
```diff
@@ -586,3 +585,86 @@ Feature: Bind an application to a service
...
+            metadata:
+                name: persisting-binding-request
+            spec:
+                services:
+                -   group: service.example.com
+                    version: v1
+                    kind: Backserv
+                    name: demo-backserv-cr-2
+            """
+        Then Get "{.metadata.resourceVersion}" of "servicebinding" "persisting-binding-request"
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:660]: what is the purpose of this step?

File: test/acceptance/features/bindAppToService.feature [hunk 15]
```diff
@@ -586,3 +585,86 @@ Feature: Bind an application to a service
...
+                services:
+                -   name: backend-demo
+            """
+        Then Error message "spec.services.group: Required value" is thrown
+        And Error message "spec.services.kind: Required value" is thrown
+        And Error message "spec.services.version: Required value" is thrown
+        And Service Binding "binding-request-without-gvk" is not persistent in the cluster
+
+        @negative1
+        Scenario: Configuring exsting service binding with invalid spec configuration is not allowed
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:646]: better reads as "Removing services field from existing serivce binding is not allowed"

File: test/acceptance/features/bindAppToService.feature (hunk 14 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:660]: This step will save the resourceVersion of first CR in context and we'll later verify if the resourceVersion is same or changed to validate the unchanged state of service binding.

File: pkg/apis/operators/v1alpha1/servicebinding_types.go (hunk 8 above)
Comment (pratikjagrut) [reply pkg/apis/operators/v1alpha1/servicebinding_types.go:117]: Yes, I tried creating SBR without spec and I was successful in doing so. Making services required will only be applicable if the spec is present in SBR if spec itself is not present then Kubernetes will not object to it.

File: test/acceptance/features/bindAppToService.feature (hunk 9 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature]: Sorry commenting it is a mistake. Reverting it back.

File: test/acceptance/features/bindAppToService.feature (hunk 10 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature]: Removed.

File: test/acceptance/features/bindAppToService.feature (hunk 15 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature]: Done.

File: pkg/apis/operators/v1alpha1/servicebinding_types.go (hunk 8 above)
Comment (pratikjagrut) [reply pkg/apis/operators/v1alpha1/servicebinding_types.go:117]: Do you feel like I should address it another PR?

Comment (pedjak) [reply pkg/apis/operators/v1alpha1/servicebinding_types.go:117]: Either that or change the scope of the PR updating its description. If you decide to keep this change, we should then add additional acceptance tests.

File: pkg/controller/servicebinding/reconciler.go [hunk 16]
```diff
@@ -78,10 +78,10 @@ func (r *reconciler) getServiceBinding(
 	return sbr, nil
 }
 
-func servicesOrEmptyList(maybeServices *[]v1alpha1.Service) []v1alpha1.Service {
+func servicesOrEmptyList(maybeServices []v1alpha1.Service) []v1alpha1.Service {
```
Comment (pedjak) [inline pkg/controller/servicebinding/reconciler.go:81]: this function is now pointless and can be removed fully, because `maybeServices` is not pointer anymore, this cannot be `nil`.

File: test/acceptance/features/bindAppToService.feature [hunk 17]
```diff
@@ -670,3 +670,86 @@ Feature: Bind an application to a service
...
+            kind: ServiceBinding
+            metadata:
+                name: binding-request-without-gvk
+            spec:
+                services:
+                -   name: backend-demo
+            """
+        Then Error message "spec.services.group: Required value" is thrown
+        And Error message "spec.services.kind: Required value" is thrown
+        And Error message "spec.services.version: Required value" is thrown
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:727]: Does user see 3 error messages? or it is just single multiline message?

File: test/acceptance/features/bindAppToService.feature (hunk 14 above)
Comment (pedjak) [reply test/acceptance/features/bindAppToService.feature:745]: This smells like abstraction leakage, i.e. implementation details. User should not know anything about that. The only thing what users sees is an error after trying to update the service binding. Think about the scenario as list of instruction we give user/tester so he is able to verify the scenario. Whatever we do in the line 732 should gather enough information (for example this step could be a part of that). so that we can perform later steps.

File: test/acceptance/features/bindAppToService.feature (hunk 17 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:727]: Single multiline error message.

Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:727]: Should I coalesce these in one multiline error step?

File: pkg/apis/operators/v1alpha1/servicebinding_types.go (hunk 8 above)
Comment (isutton) [reply pkg/apis/operators/v1alpha1/servicebinding_types.go:117]: I'd say move this to another PR.

File: pkg/controller/servicebinding/reconciler.go (hunk 16 above)
Comment (pratikjagrut) [reply pkg/controller/servicebinding/reconciler.go]: Removed.

File: test/acceptance/features/bindAppToService.feature (hunk 14 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature]: Redesigned my approach PTAL.

File: pkg/apis/operators/v1alpha1/servicebinding_types.go (hunk 8 above)
Comment (pratikjagrut) [reply pkg/apis/operators/v1alpha1/servicebinding_types.go:117]: Okay, I'll create another PR for this.

Comment (pratikjagrut) [reply pkg/apis/operators/v1alpha1/servicebinding_types.go]: Raised another PR for this: https://github.com/redhat-developer/service-binding-operator/pull/712 PTAL

File: pkg/controller/servicebinding/mapper.go [hunk 18]
```diff
@@ -42,7 +42,7 @@ func isSecret(obj runtime.Object) bool {
 
 // isSBRService checks whether the given obj is a service in given sbr.
 func isSBRService(sbr *v1alpha1.ServiceBinding, obj runtime.Object) bool {
-	services := servicesOrEmptyList(sbr.Spec.Services)
+	services := sbr.Spec.Services
```
Comment (pedjak) [inline pkg/controller/servicebinding/mapper.go:45]: do we really need `services` var? We can iterate directly `sbr.Spec.Services` in for loop?

File: test/acceptance/features/bindAppToService.feature (hunk 17 above)
Comment (pedjak) [reply test/acceptance/features/bindAppToService.feature:702]: What other messages? @pmacik wdyt?

**Review by pedjak — COMMENTED** [review]
In general looks good, need to be squashed.

File: test/acceptance/features/bindAppToService.feature (hunk 17 above)
Comment (pratikjagrut) [reply test/acceptance/features/bindAppToService.feature:702]: All the scenarios related to this PR uses the same step for error validation step, which produces single line error. So to confirm the reusability of steps I'll need to update those scenarios as well.

File: pkg/controller/servicebinding/mapper.go (hunk 18 above)
Comment (pratikjagrut) [reply pkg/controller/servicebinding/mapper.go:45]: No, we don't need it, I'll update it.

Comment (pratikjagrut) [reply pkg/controller/servicebinding/mapper.go]: Done.

**Review by pedjak — APPROVED** [review]
